src/extensions/markdown/CodeBlock/handle-paste.ts

```diff
@@ -1,52 +1,147 @@
+import type {Schema} from 'prosemirror-model';
+import type {Transaction} from 'prosemirror-state';
 import dd from 'ts-dedent';
 
 import {getLoggerFromState} from '#core';
-import {Fragment} from '#pm/model';
 import type {EditorProps} from '#pm/view';
 import {DataTransferType, isVSCode, tryParseVSCodeData} from 'src/utils/clipboard';
 
-import {CodeBlockNodeAttr} from './CodeBlockSpecs';
 import {codeBlockType} from './const';
 
+export type CodePasteData = {
+    editor: string;
+    value: string;
+    inline: boolean;
+};
+
+type InsertCodeParams = {
+    tr: Transaction;
+    schema: Schema;
+    code: CodePasteData;
+    from: number;
+    to: number;
+    inCodeBlock: boolean;
+};
+
 export const handlePaste: NonNullable<EditorProps['handlePaste']> = (view, e) => {
-    if (!e.clipboardData || view.state.selection.$from.parent.type.spec.code) return false;
-    const code = getCodeData(e.clipboardData);
+    const data = e.clipboardData;
+    if (!data) return false;
+
+    const code = getCodeData(data);
     if (!code) return false;
 
-    getLoggerFromState(view.state).event({
+    const {state} = view;
+    const {tr, schema, selection} = state;
+    const $from = selection.$from;
+    const inCodeBlock = Boolean($from.parent.type.spec.code);
+
+    logPasteEvent(state, code, data);
+
+    if (!code.value) {
+        return false;
+    }
+
+    insertCode({
+        tr,
+        schema,
+        code,
+        from: selection.from,
+        to: selection.to,
+        inCodeBlock,
+    });
+
+    view.dispatch(tr.scrollIntoView());
+    e.preventDefault();
+    return true;
+};
+
+function logPasteEvent(
```

Comment:
`logPasteEvent` is used only once, so inlining it might be simpler.

Also, the logging lost the `editorMode` field, is it ok?